elki/src/main/java/elki/math/statistics/dependence/MCDEDependenceMeasure.java

```diff
@@ -0,0 +1,249 @@
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package elki.math.statistics.dependence;
+
+import elki.utilities.datastructures.arraylike.NumberArrayAdapter;
+import elki.utilities.documentation.Reference;
+import elki.utilities.exceptions.AbortException;
+import elki.utilities.optionhandling.OptionID;
+import elki.utilities.optionhandling.Parameterizer;
+import elki.utilities.optionhandling.constraints.CommonConstraints;
+import elki.utilities.optionhandling.parameterization.Parameterization;
+import elki.utilities.optionhandling.parameters.DoubleParameter;
+import elki.utilities.optionhandling.parameters.IntParameter;
+import elki.utilities.optionhandling.parameters.ObjectParameter;
+import elki.utilities.optionhandling.parameters.RandomParameter;
+import elki.utilities.random.RandomFactory;
+import elki.math.statistics.tests.mcde.*;
+
+import java.util.Arrays;
+import java.util.Random;
+
+/**
+ * Implementation of bivariate Monte Carlo Density Estimation as described in
+ * Edouard Fouché & Klemens Böhm<br>
+ * Monte Carlo Density Estimation<br>
+ * Proc. 2019 ACM Int. Conf. on Scientific and Statistical Database Management (SSDBM 2019)
+ * <p>
+ * This is an abstract class. In order to use MCDE extend it and implement an appropriate statistical test that
+ * returns a p-value and index structure for efficient computation of the statistical test.
+ * <p>
+ * The instantiation of MCDE based on the Mann-Whitney U test is called MWP (as described in the paper).
+ *
+ * @author Alan Mazankiewicz
+ * @author Edouard Fouché
+ */
+
+@Reference(authors = "Edouard Fouché, Klemens Böhm", //
+    title = "Monte Carlo Density Estimation", //
+    booktitle = "Proc. 2019 ACM Int. Conf. on Scientific and Statistical Database Management (SSDBM 2019)", url = "https://doi.org/10.1145/3335783.3335795", //
+    bibkey = "DBLP:conf/ssdbm/FoucheB19")
+
+public class MCDEDependenceMeasure
+    extends AbstractDependenceMeasure {
+  /**
+   * Monte-Carlo iterations.
+   */
+  protected int m = 50;
+
+  /**
+   * Expected share of instances in slice (independent dimensions).
+   */
+  protected double alpha = 0.5;
+
+  /**
+   * Parameter that specifies the size of the marginal restriction. Note that in the original paper
+   * alpha = beta and as such there is no explicit distinction between the parameters.
+   */
+  protected double beta = 0.5;
+
+  /**
+   * Random generator.
+   */
+  protected RandomFactory rnd;
+
+  /**
+   * Statistical Test returning p-value tailored to MCDE Framework.
+   */
+  protected MCDETest mcdeTest;
```

Comment:
This line (and many more below) has a warning, because MCDETest has a generic.

Reply:
The easiest solution that I could find is to type this as `MCDETest<MCDETest.RankStruct>`, and perform a single cast in `MCDEDependenceMeasure`. However, this may - even as it is now - cause some subtle class cast exception when someone actually extends this later.